tests/Doctrine/Tests/DBAL/Functional/ConnectionTest.php

```diff
@@ -70,6 +70,40 @@ public function testTransactionNestingBehavior() : void
             $this->connection->rollBack();
             self::assertEquals(0, $this->connection->getTransactionNestingLevel());
         }
+
+        $this->connection->beginTransaction();
+        $this->connection->close();
+        $this->connection->beginTransaction();
+        self::assertEquals(1, $this->connection->getTransactionNestingLevel());
+    }
+
+    public function testTransactionNestingLevelIsResetOnReconnect() : void
+    {
+        if ($this->connection->getDatabasePlatform()->getName() === 'sqlite') {
+            $params           = $this->connection->getParams();
+            $params['memory'] = false;
+            $params['path']   = '/tmp/test_nesting.sqlite';
+
+            $connection = DriverManager::getConnection(
+                $params,
+                $this->connection->getConfiguration(),
+                $this->connection->getEventManager()
+            );
+        } else {
+            $connection = $this->connection;
+        }
+
+        $connection->executeQuery('CREATE TABLE test_nesting(test int not null)');
+
+        $this->connection->beginTransaction();
+        $this->connection->beginTransaction();
+        $connection->close(); // connection closed in runtime (for example if lost or another application logic)
+
+        $connection->beginTransaction(); // should connect, reset nesting level and increase it once
+        $connection->executeQuery('insert into test_nesting values (33)');
+        $connection->rollback();
+
+        self::assertEquals(0, $connection->fetchColumn('select count(*) from test_nesting'));
```

Comment:
Should the transaction nesting level be checked here too?

Reply:
Not really. The fact that the transaction was successfully rolled back is enough. The nesting level is an implementation detail in this case.

Reply:
> 
> 
> Should the transaction nesting level be checked here too?

The exact values of nestingLevel is completely tested in testTransactionNestingBehavior()